### PR TITLE
JBPM-7068 Process Definitions: Sorting by columns containing text is …

### PR DESCRIPTION
**Pull request by bxf12315**
…key sensitive

@cristianonicolai  Please review it.

### CODE REVIEW

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java [hunk 1]
```diff
@@ -261,7 +261,7 @@ protected void applySorting(List<ProcessDefinition> input, final QueryContext qu
 				@Override
 				public int compare(ProcessDefinition o1, ProcessDefinition o2) {
 					if ("ProcessName".equals(queryContext.getOrderBy())) {
-						return o1.getName().compareTo(o2.getName());
+						return o1.getName().compareToIgnoreCase(o2.getName());
 					} else if ("ProcessVersion".equals(queryContext.getOrderBy())) {
 						return o1.getVersion().compareTo(o2.getVersion());
 					} else if ("Project".equals(queryContext.getOrderBy())) {
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java]: _Project_ also needs the same fix.

**Comment by bxf12315** [comment]
@cristianonicolai  updated it.

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by cristianonicolai** [comment]
@bxf12315 fix looks ok but you still need to add tests

**Comment by cristianonicolai** [comment]
@bxf12315 you dont need to test the entire service class, you cn for instance isolate the comparator and just test that the sorting logic makes sense.

**Comment by bxf12315** [comment]
@cristianonicolai  I created test case for it .
Please review it.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java [hunk 2]
```diff
@@ -66,4 +69,20 @@ public static Long getProcessInstanceId(Command<?> command) {
 
 		return roles;
 	}
+
+	public static Comparator getIgnoreCaseComparator(final QueryContext queryContext){
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java]: @mswiderski I let you weight in here if this is the best place to have this comparator or perhaps move it to its own class.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java [hunk 3]
```diff
@@ -50,6 +56,68 @@
             new TypeAnnotationsScanner(), 
             new FieldAnnotationsScanner(), new SubTypesScanner());
 
+    private List<ProcessDefinition> expected;
+    private List<ProcessDefinition> input;
+
+    @Before
+    public void initComparatorTest(){
+        expected = new ArrayList<ProcessDefinition>() {{
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: @bxf12315 I would recommend using `Arrays.asList(...)`  instead of creating the list using inner add.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java [hunk 4]
```diff
@@ -50,6 +56,68 @@
...
+                                     null, null, "ORG.JBpm.6.service", false));
+            add(new ProcessAssetDesc(null, "c1", null, null, null,
+                                     null, null, "ORG.Jbpm.5.service", false));
+        }};
+    }
+
+    @Test
+    public void ignoreCaseComparatorUsingProjectNameTest(){
+
+        initComparatorTest();
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: I dont think using test variable is a good choice in this case, it would be best to have test isolated and just have a reusable method to get this expected and input data.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 5]
```diff
@@ -269,10 +269,10 @@ public void testGetProcessesSortByProcessName() {
     	List<String> expectedProcessIds = new ArrayList<String>();
 
     	expectedProcessIds.add("AdHoc SubProcess");
+		expectedProcessIds.add("humanTaskSample");
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: formatting?

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java [hunk 6]
```diff
@@ -50,6 +56,68 @@
...
+
+        assertEquals(expected,input);
+    }
+
+    @Test
+    public void ignoreCaseComparatorUsingProjectTest(){
+        initComparatorTest();
+        QueryContext queryContext = new QueryContext();
+        queryContext.setOrderBy("Project");
+        queryContext.setAscending(true);
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: Can you also check that `setAscending(false)` also works as expected?

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java (hunk 3 above)
Comment (bxf12315) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: ok

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 5 above)
Comment (bxf12315) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: Ok I will check it.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java (hunk 6 above)
Comment (bxf12315) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: if use setAscending(false), it is a other test case.
I will create new one

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java (hunk 2 above)
Comment (mswiderski) [reply jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java]: definitely move this to a new class - it does not make much sense to have it as anonymous class

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 5 above)
Comment (bxf12315) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: I find it is correct in my local.

**Comment by bxf12315** [comment]
@cristianonicolai @mswiderski  I was reset processDefinition Comparator code, I think we refactor it to meet the requirements of unit test. Now I find class RuntimeDataServiceImplTest.
So I just refactor two methods of RuntimeDataServiceImplTest

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java [hunk 7]
```diff
@@ -17,9 +17,12 @@
 package org.jbpm.kie.services.impl;
 
 import java.util.ArrayList;
+import java.util.Comparator;
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/CommonUtils.java]: Please revert unnecessary changes.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java [hunk 8]
```diff
@@ -16,7 +16,9 @@
 
 package org.jbpm.kie.services.impl;
 
+import static org.junit.Assert.assertEquals;
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/impl/CommonUtilsTest.java]: Please revert unnecessary changes.

File: jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java [hunk 9]
```diff
@@ -257,19 +257,18 @@ protected void applyDeploymentFilter(Map<String, Object> params) {
...
-					} else if ("Project".equals(queryContext.getOrderBy())) {
-						return o1.getDeploymentId().compareTo(o2.getDeploymentId());
-					}
-					return 0;
-				}
-			});
+                @Override
+                public int compare(ProcessDefinition o1, ProcessDefinition o2) {
+                    if ("ProcessName".equals(queryContext.getOrderBy())) {
+                        return o1.getName().compareToIgnoreCase(o2.getName());
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/main/java/org/jbpm/kie/services/impl/RuntimeDataServiceImpl.java]: I think the formatting changes in here are not necessary, would suggest to simply change the affected lines to have a simplified diff.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java (hunk 5 above)
Comment (cristianonicolai) [reply jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: From the diff, there is a clear difference between the lines above and below. For consistency, just keep the same.

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 10]
```diff
@@ -282,6 +282,28 @@ public void testGetProcessesSortByProcessName() {
     	}
     }
 
+	@Test
+	public void testGetProcessesSortByProcessNameForDsc() {
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: ForDesc?

File: jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java [hunk 11]
```diff
@@ -282,6 +282,28 @@ public void testGetProcessesSortByProcessName() {
     	}
     }
 
+	@Test
```
Comment (cristianonicolai) [inline jbpm-services/jbpm-kie-services/src/test/java/org/jbpm/kie/services/test/RuntimeDataServiceImplTest.java]: Same here regarding formatting, keep class consistent.

**Comment by bxf12315** [comment]
@cristianonicolai  It is also not correct formatting , I will try to open new PR for it.

**Comment by cristianonicolai** [comment]
@bxf12315 thanks.
